File: UsbSerialConsole/app/src/main/java/jp/sugnakys/usbserialconsole/settings/BasePreferenceFragment.java
package jp.sugnakys.usbserialconsole.settings;

import android.content.SharedPreferences;
import android.os.Bundle;

import androidx.appcompat.widget.Toolbar;
import androidx.preference.ListPreference;
import androidx.preference.Preference;
import androidx.preference.PreferenceFragmentCompat;

import java.util.Objects;

import jp.sugnakys.usbserialconsole.R;
import jp.sugnakys.usbserialconsole.util.Log;

public class BasePreferenceFragment extends PreferenceFragmentCompat implements SharedPreferences.OnSharedPreferenceChangeListener {

    private static final String TAG = "BasePreferenceFragment";

    SharedPreferences sharedPreference;

    String[] listPrefKeys;

    @Override
    public void onCreatePreferences(Bundle savedInstanceState, String rootKey) {
        // Noting to do
    }

    @Override
    public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String key) {
        Preference preference = findPreference(key);
        if (preference instanceof ListPreference) {
            ListPreference listPref = (ListPreference) preference;
            preference.setSummary(listPref.getEntry());
        }
    }

    @Override
    public void onResume() {
        super.onResume();

        sharedPreference = getPreferenceManager().getSharedPreferences();
        sharedPreference.registerOnSharedPreferenceChangeListener(this);
        setSummary();

        Toolbar toolbar = Objects.requireNonNull(getActivity()).findViewById(R.id.toolbar);
        toolbar.setNavigationOnClickListener(view -> Objects.requireNonNull(getFragmentManager()).popBackStack());
    }

    private void setSummary() {
        if (listPrefKeys == null) {
            return;
        }

        ListPreference listPref;
        for (String prefKey : listPrefKeys) {
            listPref = (ListPreference) findPreference(prefKey);
            Log.d(TAG, "Preference: " + prefKey + ", value: " + listPref.getEntry());
            listPref.setSummary(listPref.getEntry());
        }
    }

    @Override
    public void onPause() {
        sharedPreference.unregisterOnSharedPreferenceChangeListener(this);
        super.onPause();
    }
}

File: UsbSerialConsole/app/src/main/java/jp/sugnakys/usbserialconsole/settings/BasePreferenceFragment.kt
package jp.sugnakys.usbserialconsole.settings

import android.content.SharedPreferences
import android.content.SharedPreferences.OnSharedPreferenceChangeListener
import android.os.Bundle
import androidx.preference.ListPreference
import androidx.preference.Preference
import androidx.preference.PreferenceFragmentCompat
import jp.sugnakys.usbserialconsole.util.Log

open class BasePreferenceFragment : PreferenceFragmentCompat(), OnSharedPreferenceChangeListener {

    lateinit var sharedPreference: SharedPreferences

    @JvmField
    var listPrefKeys: Array<String>? = null

    companion object {
        private const val TAG = "BasePreferenceFragment"
    }

    override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
        // Noting to do
    }

    override fun onSharedPreferenceChanged(sharedPreferences: SharedPreferences, key: String) {
        val preference = findPreference<Preference>(key)
        if (preference is ListPreference) {
            preference.setSummary(preference.entry)
        }
    }

    override fun onResume() {
        super.onResume()

        sharedPreference = preferenceManager.sharedPreferences
        sharedPreference.registerOnSharedPreferenceChangeListener(this)

        setSummary()
    }

    private fun setSummary() {
        listPrefKeys?.let {
            for (prefKey in it) {
                val listPref = findPreference<ListPreference>(prefKey)
                Log.d(TAG, "Preference: " + prefKey + ", value: " + listPref?.entry)
                listPref?.summary = listPref?.entry ?: ""
            }
        }
    }

    override fun onPause() {
        sharedPreference.unregisterOnSharedPreferenceChangeListener(this)
        super.onPause()
    }
}
